File: docker/copy/update.lua
-- Set command timeout (in seconds)
local CMD_TIMEOUT = 120  -- 2 minutes per command

-- Detect if we're running in Nginx context
local is_nginx = ngx ~= nil

-- Function to safely print that works in both contexts
local function safe_print(str)
    if str then
        if is_nginx then
            ngx.print(str)
            ngx.flush(true)
        else
            print(str)
            io.flush()
        end
    end
end

-- Function to output HTML if in Nginx context, plain text otherwise
local function output_message(str, is_error)
    if is_nginx then
        if is_error then
            safe_print('<span style="color: red;">' .. str .. '</span>')
        else
            safe_print(str)
        end
    else
        if is_error then
            safe_print('ERROR: ' .. str)
        else
            safe_print(str)
        end
    end
end

local prog = require'resty.exec'.new('/tmp/exec.sock')

prog.stdout = function(v)
    safe_print(v)
end

prog.stderr = function(v)
    output_message(v, true)
end

local function run_command(...)
    local args = {...}
    safe_print('Running command: ' .. table.concat(args, ' ') .. '\n')
    
    -- Set timeout for the command
    prog.timeout = CMD_TIMEOUT * 1000  -- Convert to milliseconds
    
    local res, err = prog(...)
    if err then
        if err:match("timeout") then
            output_message('Command timed out after ' .. CMD_TIMEOUT .. ' seconds\n', true)
        else
            output_message(err .. '\n', true)
        end
        return false
    end
    return true
end

local function readAll(file)
    local f = io.open(file, "rb")
    if not f then return nil, "Could not open file: " .. file end
    local content = f:read("*all")
    f:close()
    return content and content:gsub("%s+", "") or nil, not content and "File is empty: " .. file or nil
end

-- Initialize output based on context
if is_nginx then
    ngx.header.content_type = "text/html"
    safe_print('<html><div style="white-space: pre; font-family: monospace;">')
end

-- Read spreadsheet ID at the start
local spreadsheet, err = readAll('/config/jpf.spreadsheet')
if not spreadsheet then
    output_message(err, true)
    if is_nginx then
        safe_print('</div></html>')
    end
    return
end

output_message('Starting update process...', false)
output_message('----------------------------------------', false)

-------------------------------------------------------------------------------

output_message('Copying Jekyll Template...', false)
if not run_command('mkdir','-p','/jpftemp/input') then goto error end
if not run_command('cp','-r','/website/template','/jpftemp') then goto error end

-------------------------------------------------------------------------------

output_message('----------------------------------------', false)
output_message('Downloading Google Sheet to CSV input files...', false)
output_message('Spreadsheet = '..spreadsheet, false)
if not run_command('/root/.local/bin/gs-to-csv','-f','--service-account-credential-file','/config/jpf.credentials.json',spreadsheet,'.*','/jpftemp/input') then goto error end

-------------------------------------------------------------------------------

output_message('----------------------------------------', false)
output_message('Running jpf...', false)
if not run_command('jpf','--html','/jpftemp') then goto error end

-------------------------------------------------------------------------------

output_message('----------------------------------------', false)
output_message('Copying Output...', false)
if not run_command('cp','-r','/jpftemp/output','/var/www') then goto error end

-------------------------------------------------------------------------------

output_message('----------------------------------------', false)
output_message('SUCCESS: Process completed successfully!', false)
if is_nginx then
    safe_print('</div></html>')
end
do return end

::error::
output_message('----------------------------------------', false)
output_message('ERROR: Process failed. Check the error messages above.', true)
if is_nginx then
    safe_print('</div></html>')
end

-- prog('ls','/lua')

File: docker/copy/update.py
from fastapi import FastAPI, WebSocket, WebSocketDisconnect
from fastapi.responses import FileResponse
from fastapi.middleware.cors import CORSMiddleware
from fastapi.staticfiles import StaticFiles
import subprocess
import json
import os
import time
import asyncio
import sys
from starlette.responses import Response

app = FastAPI()

# Add CORS middleware
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

# Mount static files
#app.mount("/static", StaticFiles(directory="/website/static"), name="static")
#app.mount("/output", StaticFiles(directory="/var/www/output/html"), name="output")

# Define all commands to be run
COMMANDS = [
    {
        "message": "Setting up working directory...",
        "commands": [
            {"cmd": ["mkdir", "-p", "/jpftemp/input"], "type": "run"},
            {"cmd": ["cp", "-r", "/website/template", "/jpftemp"], "type": "run"}
        ]
    },
    {
        "message": "Downloading Google Sheet to CSV input files...",
        "commands": [
            {
                "cmd": ["/root/.local/bin/gs-to-csv", "-f", "--service-account-credential-file", 
                       "/config/jpf.credentials.json", "{spreadsheet_id}", ".*", "/jpftemp/input"]
            }
        ]
    },
    {
        "message": "Running JPF to generate website...",
        "commands": [
            {"cmd": ["jpf", "--html", "/jpftemp"]}
        ]
    },
    {
        "message": "Copying generated website to web root...",
        "commands": [
            {"cmd": ["cp", "-r", "/jpftemp/output", "/var/www"], "type": "run"}
        ]
    }
]

def read_spreadsheet_id():
    """Read the Google Sheets ID from the configuration file."""
    try:
        with open('/config/jpf.spreadsheet', 'r') as f:
            return f.read().strip()
    except Exception as e:
        return None

async def send_message(websocket: WebSocket, message: str, color: str, close: bool = False):
    """Send a message to the client via WebSocket."""
    data = {
        'output': message,
        'color': color
    }
    if close:
        data['close'] = True
    await websocket.send_json(data)
    # Remove the empty text message that's causing JSON parse errors
    # await websocket.send_text("")

async def stream_command_output(command, websocket: WebSocket):
    """Execute a command and stream its output in real-time."""
    process = await asyncio.create_subprocess_exec(
        *command,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
        env={**os.environ, 'PYTHONUNBUFFERED': '1'}
    )

    async def read_stream(stream, color):
        while True:
            line = await stream.readline()
            if not line:
                break
            line = line.decode().strip()
            if line:
                await send_message(websocket, line, color)

    # Read stdout and stderr concurrently
    await asyncio.gather(
        read_stream(process.stdout, 'black'),
        read_stream(process.stderr, 'red')
    )

    # Wait for the process to complete
    return_code = await process.wait()
    if return_code != 0:
        await send_message(websocket, f"Command failed with return code {return_code}", 'red')

@app.get("/update")
async def update():
    """Serve the update.html page with no-cache headers."""
    response = FileResponse("/website/static/update.html")
    response.headers["Cache-Control"] = "no-store, no-cache, must-revalidate, max-age=0"
    response.headers["Pragma"] = "no-cache"
    response.headers["Expires"] = "0"
    return response

@app.websocket("/update/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    
    try:
        # Read spreadsheet ID
        spreadsheet = read_spreadsheet_id()
        if not spreadsheet:
            await send_message(websocket, 'Could not read spreadsheet ID', 'red')
            await send_message(websocket, '<br>Connection closed due to error.', 'red', close=True)
            return

        # Execute all commands in sequence
        for step in COMMANDS:
            await send_message(websocket, step['message'], 'blue')
            
            for command in step['commands']:
                # Replace spreadsheet_id placeholder if present
                cmd = [arg.format(spreadsheet_id=spreadsheet) if '{spreadsheet_id}' in arg else arg 
                      for arg in command['cmd']]
                
                if command.get('type') == 'run':
                    try:
                        process = await asyncio.create_subprocess_exec(
                            *cmd,
                            stdout=asyncio.subprocess.PIPE,
                            stderr=asyncio.subprocess.PIPE
                        )
                        await process.wait()
                        if process.returncode != 0:
                            raise subprocess.CalledProcessError(process.returncode, cmd)
                    except subprocess.CalledProcessError as e:
                        await send_message(websocket, f'Command failed: {e}', 'red')
                        await send_message(websocket, '<br>Connection closed due to error.', 'red', close=True)
                        return
                else:  # default to stream
                    await stream_command_output(cmd, websocket)

        await send_message(websocket, 'SUCCESS: Website has been regenerated successfully!', 'green')
        await send_message(websocket, '<br>All done.', 'green', close=True)

    except WebSocketDisconnect:
        print("Client disconnected")
    except Exception as e:
        print(f"Error: {e}")
        try:
            await send_message(websocket, f'Error: {str(e)}', 'red')
            await send_message(websocket, '<br>Connection closed due to error.', 'red', close=True)
        except:
            pass

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000) 